### PR TITLE
babashka: use upstream version of Clojure tools

### PR DESCRIPTION
**Pull request by DerGuteMoritz**
## Description of changes

As recommended by @borkdude[1], we should use the Clojure tools version included in a particular upstream release since the code might not always work with more recent versions.

[1]  https://github.com/borkdude/deps.clj/issues/113#issuecomment-1735805109

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

- - -
This change is [Reviewable](https://reviewable.io/reviews/NixOS/nixpkgs/257473)

### CODE REVIEW

File: pkgs/development/interpreters/babashka/wrapped.nix
Comment (SuperSandro2000) [inline pkgs/development/interpreters/babashka/wrapped.nix:40]: That is probably triggering IFD which would make ofborg fail

**Comment by SuperSandro2000** [comment]
> As recommended by @borkdude[1], we should use the Clojure tools version included in a particular upstream release since the code might not always work with more recent versions.

We don't want to vendor things. We should rather use an appropriate closure tools version from nixpkgs or add an override.

Also can we add a short test to know if things break on updates? Than we would be confident that things work even if we don't match the upstream version.

**Review by jlesquembre — COMMENTED** [review]
It makes sense to try to use the recommended version, but in that case, we should add an script to automate the updates and keep the version in sync.
With the current approach, next time babashka updates clojure tools, the build will break.

File: pkgs/development/interpreters/babashka/wrapped.nix
Comment (DerGuteMoritz) [reply pkgs/development/interpreters/babashka/wrapped.nix:21]: Thanks for the pointer!

File: pkgs/development/interpreters/babashka/wrapped.nix
Comment (DerGuteMoritz) [reply pkgs/development/interpreters/babashka/wrapped.nix:40]: Interesting! What in particular about this expression is causing this and where could I read up on IFD?

Comment (DerGuteMoritz) [reply pkgs/development/interpreters/babashka/wrapped.nix:40]: Ah, didn't occur to me to check the inside jar :bulb: Much nicer, thanks!

File: pkgs/development/interpreters/babashka/wrapped.nix
Comment (DerGuteMoritz) [reply pkgs/development/interpreters/babashka/wrapped.nix:46]: I think I'll switch this to using the `clojure` derivation instead (as suggested by https://github.com/NixOS/nixpkgs/pull/257473#issuecomment-1735958095), with the version overridden accordingly.

File: pkgs/development/interpreters/babashka/wrapped.nix
Comment (DerGuteMoritz) [reply pkgs/development/interpreters/babashka/wrapped.nix:69]: Hmm I thought my embedding of `${babashka-unwrapped.installCheckPhase}` did exactly that? I agree that using `finalAttrs.installCheckPhase` would be cleaner and I'll switch to that. But in your suggestion you prepend that and *also* keep the `${babashka-unwrapped.installCheckPhase}` which probably wasn't intentional, was it?

**Comment by DerGuteMoritz** [comment]
> We don't want to vendor things. We should rather use an appropriate closure tools version from nixpkgs or add an override.

@SuperSandro2000 Makes sense! I'll switch it over to using the `clojure` derivation as before and override its version to the expected one. Should have thought of that!

> Also can we add a short test to know if things break on updates? Than we would be confident that things work even if we don't match the upstream version.

I think the best we can do is detect whether we need to update the tracked version on our end. Testing whether it works with a non-matching version seems intractable as the surface area of clojure tools is quite large and fully exposed via `bb clojure`. Will look into the former!

**Comment by DerGuteMoritz** [comment]
> It makes sense to try to use the recommended version, but in that case, we should add an script to automate the updates and keep the version in sync. With the current approach, next time babashka updates clojure tools, the build will break.

@jlesquembre Good call, that's a better approach overall. I'll rework the patch accordingly (might take a few days, tho).

File: pkgs/development/interpreters/babashka/wrapped.nix
Comment (jlesquembre) [reply pkgs/development/interpreters/babashka/wrapped.nix:69]: ups, sorry, I missed the `${babashka-unwrapped.installCheckPhase}` part. Ignore my comment. 

In fact, I like more your approach of using  `${babashka-unwrapped.installCheckPhase}`, it makes clear that we are running the same tests again.  But as long as we run the same tests again, I don't have a strong preference on `babashka-unwrapped.installCheckPhase` vs `finalAttrs.installCheckPhase`, up to you.

File: pkgs/development/interpreters/babashka/wrapped.nix
Comment (jlesquembre) [reply pkgs/development/interpreters/babashka/wrapped.nix:46]: Makes sense. I'm not sure how an override will look here, looking forward to seeing what you have in mind.

**Comment by jlesquembre** [comment]
> Good call, that's a better approach overall. I'll rework the patch accordingly (might take a few days, tho).

@DerGuteMoritz thanks for working on it :)

**Comment by DerGuteMoritz** [comment]
@jlesquembre OK got back to it earlier than expected :smile: Check it out! 

Note that `updateScript` was broken since the introduction of the wrapper: it still invoked `update-source-version babashka` but it should now have invoked `update-source-version babashka-unwrapped` instead! Fixed while I was in there anyway.

Hope the changes make sense and be sure to squash before merge (not sure if GH automatically squashes fixup commits?)

**Comment by DerGuteMoritz** [comment]
Forgot to mention that the `bb clojure --version` invocation in `installCheckPhase` will fail if the provided clojure version doesn't match the version expected by babashka. This is because t will then try to auto download the expected version and fail because it may not connect to the internet at that point.

File: pkgs/development/interpreters/babashka/default.nix
Comment (DerGuteMoritz) [inline pkgs/development/interpreters/babashka/default.nix:57]: Ah, I should probably wrap that `$(...)` expression in double quotes, too. Will do with the next round of addressing review feedback!

**Comment by jlesquembre** [comment]
Thanks @DerGuteMoritz , looks good. Before approving, just some minor suggestions:
- Could you format the `clojure.nix` file with nixpkgs-fmt?
- I think we should rename `clojure.nix` to `babashka-clojure.nix`, `babashka-clojure-tools.nix` or something similar, to make even clearer why we override clojure. Also, a comment about it and a link to https://github.com/borkdude/deps.clj#deps_clj_tools_version in `clojure.nix` or `wrapped.nix` would be nice.
- Could you rebase and force push? I don't think nixpkgs repo has enabled the "squash and merge" feature.


@SuperSandro2000 Could you confirm if that is what you had in mind about using an override? To me it looks good, but I want to confirm it.

File: pkgs/development/interpreters/babashka/default.nix
Comment (jlesquembre) [reply pkgs/development/interpreters/babashka/default.nix:57]: I tested the script locally and it works :) But you are right, better to wrap the expression

**Comment by DerGuteMoritz** [comment]
Hi @jlesquembre, thanks for your feedback which I have now addressed as follows:

> * Could you format the `clojure.nix` file with nixpkgs-fmt?

TIL & done!

> * I think we should rename `clojure.nix` to `babashka-clojure.nix`, `babashka-clojure-tools.nix` or something similar, to make even clearer why we override clojure. Also, a comment about it and a link to https://github.com/borkdude/deps.clj#deps_clj_tools_version in `clojure.nix` or `wrapped.nix` would be nice.

I went to rename it to `clojure-tools.nix` instead, the directory already scopes it well enough to be babashka-specific IMHO. Also changed all references to "clojure" within the code to "clojure-tools" - while it expects a `clojure`-like derivation, it's really the tools we're interested in and since that's also the term used by babashka itself, I think that makes sense.

Instead of adding comments to `clojure.nix` and `wrapped.nix`, I added a comment to `clojure-tools.nix` explaining its purpose. I think readers wondering what that thing is about will be able to find it there and that way it's not duplicated.

> * Could you rebase and force push? I don't think nixpkgs repo has enabled the "squash and merge" feature.

Done that, too!

However, there's more: 

* Remember how I originally extracted `DEPS_CLJ_RELEASED_VERSION` from the standalone jar to determine the Clojure tools version? Well, I had a quick chat with @borkdude about it and it turns out that that's *not* quite the right place to look for it. As the name suggests, it really is the version of [`deps.clj`](https://github.com/borkdude/deps.clj) which usually matches the version of the Clojure tools releases but not necessarily so. For example, for the very recently released version 1.3.185, the old command yields:
  ```
  $ curl -fsL "https://github.com/babashka/babashka/releases/download/v1.3.185/babashka-1.3.185-standalone.jar" | bsdtar -qxOf - DEPS_CLJ_RELEASED_VERSION
  1.11.1.1403
  ```
  But if you look at [the bump of `deps.clj`](https://github.com/babashka/babashka/commit/5cab060bbba8e0648e8f88d4cf4a36b35b0cd2cf)  included in that release, you'll notice that the changelog entry mentions "Catch up with CLI `1.11.1.1413`" (which is what prompted my investigation). 

  Instead, the correct place to get the version from is [this definition](https://github.com/borkdude/deps.clj/blob/cc9ddf48625dbc29943a3f3fa4b9c54e19622eec/src/borkdude/deps.clj#L17-L19). Luckily, that file is also included in source form in the jar, so I changed to update script to extract that instead and parse it using babashka itself. That means I had to wrap the whole derivation in a `let ...` to be able to do that, so diff now best viewed with whitespace ignored.
* While working on the aforementioned change, I noticed that the error handling in the update script didn't quite behave as I thought it would: Even when the version extraction code failed, it would proceed to run `update-source-version babashka.clojure-tools ...` with an empty version string. After some headscratching I remembered that command substitutions such as `$(...)` don't inherit the parent shell's options, crucially `set -e`, so it would happily march on as if nothing happened. After some digging I found `shopt -s inherit_errexit` which *does* make it inherit that option. However, it still didn't seem to behave any different. Puzzled, I trawled through the web for help and ended up [here](https://mywiki.wooledge.org/BashFAQ/105#line-204): 
  >  Another pitfall associated with `set -e` occurs when you use commands that look like assignments but aren't, such as `export`, `declare`, `typeset` or `local`.
  > [...]
  > [The exit status of the command substituion] won't trigger the `set -e` because the exit status of `local` masks it (the assignment to the variable succeeds, so `local` returns status 0).
  
  While not mentioned here, it appears that this also applies to `readonly` :grimacing: Not sure why it was used here in the first place but it doesn't seem to be overly important so I decided to just rip it out and voilà, errors now terminate the script even when something goes wrong in these command substitutions :sweat_drops: 
  
  Maybe we should just rewrite the whole update script in babashka :smile: 

Anyway, hope this all makes sense. I'm afraid another careful review is necessary now, though :pray:

**Comment by jlesquembre** [comment]
@DerGuteMoritz thanks, just 2 things I noticed: 
- Is there a reason for  `let babashka-unwrapped = buildGraalvmNativeImage {...} in babashka-unwrapped ` instead of just ` buildGraalvmNativeImage {...}`?
- You use babashka itself in the update script (which I like :-) ), in that case, I just want to mention that maybe be could replace more bash with babashka (If that makes sense to you, I didn't look in too much detail to the new update script yet)

Thanks again, I'll review it in more details in the upcoming days

**Comment by DerGuteMoritz** [comment]
> -   Is there a reason for `let babashka-unwrapped = buildGraalvmNativeImage {...} in babashka-unwrapped` instead of just `buildGraalvmNativeImage {...}`?

I think you answered the question yourself in the next item:

> -   You use babashka itself in the update script 

i.e. the reason is so that I can do that :slightly_smiling_face:  Or is there a nicer way?

As for rewriting the whole script in babashka: I'll give it a shot once this PR has landed. Was a deep enough rabbit hole as it is already and I'm sure that this would be an even deeper dig :smile:

**Comment by jlesquembre** [comment]
@DerGuteMoritz I did another test, a looks good, thanks :)

Regarding the `let babashka-unwrapped = buildGraalvmNativeImage {...}`, I don't see a better way. I'd like to update `buildGraalvmNativeImage` to mimic the behavior of `mkDerivation`, so we could do something like  `buildGraalvmNativeImage (finalAttrs: ...)`, and remove the `let` binding and the `rec`, but that's something for the future. I'm adding it to my TODO list, but no promises.

Regarding rewriting the script in babashaka, I agree, better to do it on another PR. I want to mention that I have plans to work on some helpers to make easier to integrate babashka in nix derivations. Something similar to https://github.com/DeterminateSystems/nuenv but with babashka, I'd like to write complex derivations with Nix + Babahska, without any bash. Probably I'll start with it on https://github.com/jlesquembre/clj-nix, but if it's successful, I'd like to upstream it to nixpkgs.

**Review by jlesquembre — APPROVED** [review]

**Review by thiagokokada — COMMENTED** [review]
Builds are failing in darwin.

File: pkgs/development/interpreters/babashka/default.nix
Comment (jlesquembre) [inline pkgs/development/interpreters/babashka/default.nix:35]:
```suggestion
    installCheckPhase = ''
      export CLJ_CONFIG="$TMP/.clojure"
      $out/bin/bb --version | grep '${version}'
```

That should fix the darwin issue

Comment (DerGuteMoritz) [reply pkgs/development/interpreters/babashka/default.nix:35]: Hmm why would that only happen on Darwin? Feeling a bit reluctant to apply the fix without understanding that first... any clue?

**Comment by DerGuteMoritz** [comment]
@jlesquembre 

> Regarding the `let babashka-unwrapped = buildGraalvmNativeImage {...}`, I don't see a better way. I'd like to update `buildGraalvmNativeImage` to mimic the behavior of `mkDerivation`, so we could do something like `buildGraalvmNativeImage (finalAttrs: ...)`, and remove the `let` binding and the `rec`, but that's something for the future. I'm adding it to my TODO list, but no promises.

Cool, sounds good! OK then, let's keep it as it is for now.

> Regarding rewriting the script in babashaka, I agree, better to do it on another PR.

:+1: 

> I want to mention that I have plans to work on some helpers to make easier to integrate babashka in nix derivations. Something similar to [https://github.com/DeterminateSystems/nuenv](https://github.com/DeterminateSystems/nuenv) but with babashka, I'd like to write complex derivations with Nix + Babahska, without any bash. Probably I'll start with it on [https://github.com/jlesquembre/clj-nix](https://github.com/jlesquembre/clj-nix), but if it's successful, I'd like to upstream it to nixpkgs.

That would be excellent. Ping me when you have something to try out, then I can give the conversion of the babashka package a go and hopefully provide you with some feedback :slightly_smiling_face:

File: pkgs/development/interpreters/babashka/default.nix
Comment (DerGuteMoritz) [reply pkgs/development/interpreters/babashka/default.nix:35]: (unfortunately, I don't have access to a macOS to poke at it myself ...)

Comment (jlesquembre) [reply pkgs/development/interpreters/babashka/default.nix:35]: yep, sorry for not explaining the reason.  

In the test phase we invoke `bb clojure --version`, which mimics Clojure CLI.  Clojure determines the path to the Clojure user configuration (e.g.: `~/.clojure`), and checks for a `deps.edn` file on that location. If there isn't any, it copies the default `deps.edn` file there. On Linux, in the Nix build sandbox, the path is `/build/.clojure/deps.edn`, which is writable. But on MacOS is `/var/empty/.clojure/deps.edn`, and the process fails because it can't write there. 

We can set `CLJ_CONFIG` to use a different path, so the process can copy the `deps.edn` file.

I wrote a gist with a deeper explanation:
https://gist.github.com/jlesquembre/ad7cb8357fe13312eb23a437b87bcd2b

File: pkgs/development/interpreters/babashka/wrapped.nix
Comment (SuperSandro2000) [reply pkgs/development/interpreters/babashka/wrapped.nix]: You are trying to read a file into nix code that is from a fixed output derivation.

Maybe https://nixos.wiki/wiki/Import_From_Derivation ?

File: pkgs/development/interpreters/babashka/default.nix
Comment (DerGuteMoritz) [reply pkgs/development/interpreters/babashka/default.nix:35]: Wow, thanks a lot for that deep dive! :bow: I took the liberty to file [an issue for `deps.clj` linking to your write up](https://github.com/borkdude/deps.clj/issues/114) and added your workaround with a comment pointing to it to the code.

While in there, I also hardened the version test by quoting the expected version and passing the `w` and `F` flags to `grep` so that we match the version exactly. Without that,  "1.11.1.1413" would also have satisfied versions which happen to be substrings of that version, e.g. "1.11.1.14". 

I think with that we're finally good to merge (provided the tests pass this time ...)

Comment (DerGuteMoritz) [reply pkgs/development/interpreters/babashka/default.nix:35]: Addendum: The version check is also hardened in another way in that the dots aren't interpreted as "any character" anymore which would also have caused ambiguity (that's what the `F` flag is for).

File: pkgs/development/interpreters/babashka/wrapped.nix
Comment (thiagokokada) [inline pkgs/development/interpreters/babashka/wrapped.nix:51]: This check looks like it broke the @ofborg build, can you remove it?

Comment (thiagokokada) [reply pkgs/development/interpreters/babashka/wrapped.nix:51]: Interesting, I would add this to the `buildGraalvmNativeImage` instead of just babashka (and for all environments, not just `x86_64-darwin`).

Comment (jlesquembre) [reply pkgs/development/interpreters/babashka/wrapped.nix:51]: According to https://www.graalvm.org/latest/reference-manual/native-image/overview/BuildOptions/ , the `-march` option impacts the performance of the native executable. I agree that makes sense to add it to  `buildGraalvmNativeImage`, but I wouldn't add to all environments.

It could be interesting look at Babashka performance with different `march` values.

Comment (DerGuteMoritz) [reply pkgs/development/interpreters/babashka/wrapped.nix:51]: That makes sense. How about adding it in a way so that it's still overridable if desired?

Comment (DerGuteMoritz) [reply pkgs/development/interpreters/babashka/wrapped.nix:51]: Nevermind, I'll just add it to the default `nativeImageBuildArgs` which sufficiently supports that case I'd say.

Comment (DerGuteMoritz) [reply pkgs/development/interpreters/babashka/wrapped.nix:51]: Made a separate commit for that change - or would it be preferable to break it out into a separate PR even?

Comment (thiagokokada) [reply pkgs/development/interpreters/babashka/wrapped.nix:51]: A separate commit is fine.

File: pkgs/build-support/build-graalvm-native-image/default.nix
Comment (thiagokokada) [inline pkgs/build-support/build-graalvm-native-image/default.nix:16]:
```suggestion
    "-march=compatibility"
```

Comment (DerGuteMoritz) [reply pkgs/build-support/build-graalvm-native-image/default.nix]: Ah damn, I copied & pasted it from https://github.com/NixOS/nixpkgs/pull/257473#discussion_r1355216818 and didn't notice the typo! How did you? The tests didn't fail after all :thinking: Anyway, fixed now. Also rebased the branch once more for good measure.

Comment (thiagokokada) [reply pkgs/build-support/build-graalvm-native-image/default.nix]: Ah I see, I think you didn't realize that we run the builds in our CI.

If you look into the checks, you will eventually see (after `ofborg-eval` check finishes) something like:

```
babashka, babashka.passthru.tests on x86_64-linux
```

You can click in `Details` to see the build, and if you click you also go to the OfBorg Viewer interface to see more details. Example of the build with the typo:
https://logs.ofborg.org/?key=nixos/nixpkgs.257473&attempt_id=f7c7e3ce-6661-4411-ac5b-d37870f402e1

We run checks for 4 platforms: `aarch64-linux`, `x86_64-linux`, `aarch64-darwin` and `x86_64-linux`. So even if you don't have a macOS, you can see the build failures directly in the CI (it will take a while though, because the macOS CI is way slower than Linux CI).

In case it doesn't trigger the build step automatically (we extract which package to build from the commit messages, this is one of the reasons that commit messages are important BTW), you can trigger it manually by running `@ofborg build <package>`.

See more documentation about OfBorg here: https://github.com/NixOS/ofborg

Comment (DerGuteMoritz) [reply pkgs/build-support/build-graalvm-native-image/default.nix]: Thanks for the info, much appreciated :pray: I roughly knew that that's happening but since the overall checks status was "success", I didn't bother to check. Looks like these test jobs are configured to be allowed to fail? But it still makes sense to check them when they do I guess! *makes mental note*

Anyway, hope we're good to go now finally :sweat_drops:

Comment (thiagokokada) [reply pkgs/build-support/build-graalvm-native-image/default.nix]: > Looks like these test jobs are configured to be allowed to fail? But it still makes sense to check them when they do I guess! _makes mental note_

Yes, we allow the checks to fail because some packages are expect to fail (e.g.: maybe the package is broken in macOS, but works in Linux). Also some packages are broken in `master` branch for some reason, and we shouldn't break builds just because a package is updated (but still broken).

**Review by thiagokokada — APPROVED** [review]

**Comment by jlesquembre** [comment]
Thanks @DerGuteMoritz and @thiagokokada . This PR took long, but I'm happy with the outcome, we fixed a couple of issues, including a fix to upstream Babahska :tada:

**Comment by jlesquembre** [comment]
@DerGuteMoritz the `updateScript` is run automatically by https://github.com/r-ryantm bot:
- Docs: https://ryantm.github.io/nixpkgs-update/
- Logs: https://r.ryantm.com/log/

As far as I know, it's executed in a best-effort basis, meaning that sometimes you have to wait longer. As far as I remember, usually updates packages 2-3 days after a release, but sometimes I had to wait for a week, or even a bit longer.

**Pull request merged**
